src/sage/rings/polynomial/q_integer_valued_polynomials.py

```diff
@@ -0,0 +1,1170 @@
+# ***************************************************************************
+from sage.arith.misc import binomial
+from sage.categories.algebras import Algebras
+from sage.categories.realizations import Category_realization_of_parent
+from sage.categories.rings import Rings
+from sage.combinat.free_module import CombinatorialFreeModule
+from sage.combinat.q_analogues import q_binomial, q_int
+from sage.data_structures.blas_dict import linear_combination
+from sage.matrix.constructor import matrix
+from sage.misc.bindable_class import BindableClass
+from sage.misc.cachefunc import cached_method
+from sage.modules.free_module_element import vector
+from sage.rings.integer_ring import ZZ
+from sage.rings.polynomial.laurent_polynomial_ring import LaurentPolynomialRing
+from sage.rings.polynomial.polynomial_ring import polygen
+from sage.rings.polynomial.polynomial_ring_constructor import PolynomialRing
+from sage.rings.rational_field import QQ
+from sage.sets.family import Family
+from sage.sets.non_negative_integers import NonNegativeIntegers
+from sage.structure.parent import Parent
+from sage.structure.unique_representation import UniqueRepresentation
+
+
+def q_int_x(n):
+    """
+    Return the interpolating polynomial of `q`-integers.
+
+    INPUT:
+
+    - ``n`` -- a positive integer
+
+    EXAMPLES::
+
+        sage: from sage.rings.polynomial.q_integer_valued_polynomials import q_int_x
+        sage: q_int_x(3)
+        q^2*x + q + 1
+    """
+    ring_q = PolynomialRing(ZZ, 'q')
+    q = ring_q.gen()
+    x = polygen(ring_q, 'x')
+    return q_int(n - 1) + q**(n - 1) * x
+
+
+def q_binomial_x(m, n):
+    r"""
+    Return a `q`-analogue of ``binomial(m + x, n)``.
+
+    When evaluated at the `q`-integer `[k]_q`, this gives
+    the usual `q`-binomial coefficient `[m + k, n]_q`.
+
+    EXAMPLES::
+
+        sage: from sage.combinat.q_analogues import q_int
+        sage: from sage.rings.polynomial.q_integer_valued_polynomials import q_binomial_x, q_int_x
+        sage: q_binomial_x(4,2)(0) == q_binomial(4,2)
+        True
+        sage: q_binomial_x(3,2)(1) == q_binomial(4,2)
+        True
+        sage: q_binomial_x(3,1) == q_int_x(4)
+        True
+        sage: q_binomial_x(2,0).parent()
+        Univariate Polynomial Ring in x over Fraction Field of
+        Univariate Polynomial Ring in q over Integer Ring
+    """
+    ring = PolynomialRing(PolynomialRing(ZZ, 'q').fraction_field(), 'x')
+    if n == 0:
+        return ring.one()
+    return ring.prod(q_int_x(m + 2 - i) / q_int(i) for i in range(1, n + 1))
+
+
+class QuantumValuedPolynomialRing(UniqueRepresentation, Parent):
+    r"""
+    The quantum-valued polynomial ring on some generators over a base ring.
+
+    Quantum-valued polynomial rings are commutative and associative
+    algebras, with a basis indexed by integers.
+
+    This is endowed with two bases, named ``B`` or ``Binomial``
+    and ``S`` or ``Shifted``.
+
+    INPUT:
+
+    - ``R`` -- commutative ring
```

Comment:
I think it would be good to have `q` be included as an argument, subsequently it would be a part of the ground ring. Now you can have an option like if the input is :
1. Only `R`, then append `q` as currently implemented.
2. Both `(R, q)`, then do `q = R(q)`.
3. Just `q` and get `R = q.parent()`.

Reply:
Hmm, I would rather not allow a custom q. The ring R is not there to provide q, in my mind. I have added a sentence about that in the doc.

Reply:
I am not sure. For my own usage, I have a base ring of coefficients, almost always QQ, and the variable q on top of that. For this algebra of quantum-valued polynomials, I can see no interest to have q a complex number. Of course, one can evaluate to polynomials and then evaluate q if one will.

Reply:
I believe certain Askey scheme $q$-orthogonal polynomials (e.g., [cont. q-Hermite](https://en.wikipedia.org/wiki/Continuous_q-Hermite_polynomials)) would be in this ring, and problems I've worked on, we want to consider $q \in \mathbb{R}_{>0}$. There might also be important considerations for doing things over finite fields (or $q$ satisfying some relation). I think it would be better to be more general as it is easy to add and keep the most common behavior ($q$ is a generic parameter).

Reply:
I have now allowed to provide an optional q. Then the ring must contain this q.